### PR TITLE
[FIRRTL] Composable LowerAnnotations

### PR DESCRIPTION
**Pull request by seldridge**
Make modifications to enable `LowerAnnotations` to be run serially with normal annotation processing.  This is intended to enable an annotation-by-annotation migration path for annotations without requiring the massive changes of something like #2170.

This does three things:

1. The `--new-anno` option (which enables `LowerAnnotations`) now has no effect on the parser.  The alternate code path in the parser that this enabled (to not apply annotations to created operations) is removed.
2. `LowerAnnotations` now will only process annotations in a non-standard attribute on the circuit keyed by `annotationsForLowerAnnotations`.  
3. The parser is modified to know about the "annotationsForLowerAnnotations" key (coming from FIRAnnotations) and to put any annotations behind this key onto the circuit with the same key.

The work of (2) and (3) are only temporary while we deal with migrating annotations from FIRParser/FIRAnnotations into the `LowerAnnotations` pass.

This PR is part of work to enable Grand Central to not use `DontTouchAnnotation`s (represented as public symbols that block optimizations).  Specifically, it is easiest to change Grand Central annotation handling inside `LowerAnnotations` to attach a _private_ symbol and to use the `hw::InnerRefAttr` associated with this private symbol instead of its current numeric target ID.  This PR enables only making this migration for certain Grand Central annotations without having to do the full work of #2170.  This also enables much more parallel work where annotations can be moved one-by-one and by many people simultaneously.

### CODE REVIEW

**Comment by seldridge** [comment]
I updated this PR to use the non-standard attribute key `rawAnnotations` instead of the verbose `annotationsForLowerAnnotations` per @darthscsi suggestion.

File: include/circt/Dialect/FIRRTL/AnnotationDetails.h [hunk 1]
```diff
@@ -18,6 +18,12 @@
 namespace circt {
 namespace firrtl {
 
+//===----------------------------------------------------------------------===//
+// Common strings related to annotations
+//===----------------------------------------------------------------------===//
+
+constexpr const char *rawAnnotations = "rawAnnotations";
```
Comment (darthscsi) [inline include/circt/Dialect/FIRRTL/AnnotationDetails.h:25]: nit constexpr char[] = ...;

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp [hunk 2]
```diff
@@ -2809,24 +2809,20 @@ ParseResult FIRStmtParser::parseInstance() {
   }
 
   InstanceOp result;
-  if (getConstants().options.rawAnnotations) {
```
Comment (darthscsi) [inline lib/Dialect/FIRRTL/Import/FIRParser.cpp:2812]: Shouldn't the parser option go away?

File: include/circt/Dialect/FIRRTL/AnnotationDetails.h (hunk 1 above)
Comment (seldridge) [reply include/circt/Dialect/FIRRTL/AnnotationDetails.h:25]: I won't change that here given that this is consistent with the existing `char *` usage in the rest of this file. Fine to swing this after the fact.

**Review by darthscsi — COMMENTED** [review]
I the rawAnnotations seems to be expected in the json files.
I was envisioning:
1) parse json
2) for each annotation, query the class against the table of defered annotations and either put in the "attach to circuit in parser" bucket or the "put in rawAnnotation" attr bucket.
3) proceed as normal.

File: lib/Dialect/FIRRTL/Import/FIRParser.cpp (hunk 2 above)
Comment (seldridge) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:2812]: Yes, this should be totally gone now. 👍

Comment (seldridge) [reply lib/Dialect/FIRRTL/Import/FIRParser.cpp:2812]: I removed `rawAnnotations` from the parser options in the force-push.

File: include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h [hunk 3]
```diff
@@ -75,6 +76,253 @@ Optional<AnnoPathValue> resolveEntities(TokenAnnoTarget path, CircuitOp circuit,
...
+  auto target = anno.getNamed("target");
+  if (target)
+    return stdResolveImpl(target->getValue().cast<StringAttr>().getValue(),
+                          state);
+  return AnnoPathValue(state.circuit);
+}
+
+} // namespace
+
+const llvm::StringMap<AnnoRecord> annotationRecords{{
```
Comment (darthscsi) [inline include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h:317]: You don't want to pull all this stuff into a header (linking and namespace issues).  Can't you just defined a
`bool isAnnoClassLowered(StringAttr class);`
as the api between LowerAnno and FIRParser?

Comment (seldridge) [reply include/circt/Dialect/FIRRTL/FIRRTLAnnotationHelper.h]: Cleaned up in the force-push. Awesome suggestion.

**Review by darthscsi — APPROVED** [review]

**Pull request merged**